src/versionUtils.ts

```diff
@@ -0,0 +1,159 @@
+		if (runtimeVersionSetting && runtimeVersionSetting.toLowerCase() !== versionToUse.toLowerCase()) {
+			versionToUse = runtimeVersionSetting;
+			extension.shareMessageInMainOutputChannel(`Using Apache Camel K CLI version ${versionToUse} instead of latest version ${version}`);
+		}
+
+		if (versionToUse) {
+			await checkKamelCLIVersion().then((currentVersion) => {
+				if (versionToUse && versionToUse.toLowerCase() === currentVersion.toLowerCase()) {
+					// no need to install, it's already here
+					resolve(false);
+					return false;
+				} else {
+					resolve(true);
+					return true;
+				}
+			}).catch ( (error) => {
+				console.error(error);
+				resolve(true);
+				return true;
+			});
+		}
+	});
+}
+
+export async function pingGithubUrl(urlStr: string) : Promise<boolean> {
+	// placeholder to ensure that this url path is accessible
+	console.log(`Validate that ${urlStr} is accessible`);
+	try {
+		const res = await fetch(urlStr);
+		if (res.status === 200) {
+			return Promise.resolve(true);
+		}
+		return Promise.resolve(false);
+	} catch (err) {
+		return Promise.resolve(false);
+	}
+}
+
+export async function getLatestCamelKVersion(): Promise<Errorable<string>> {
+	const latestURL = 'https://api.github.com/repos/apache/camel-k/releases/latest';
+	const downloadResult = await downloader.toTempFile(latestURL);
+	if (failed(downloadResult)) {
+		return { succeeded: false, error: [`Failed to establish Apache Camel K stable version: ${downloadResult.error[0]}`] };
+	}
+	const rawtext = fs.readFileSync(downloadResult.result, 'utf-8');
+	let latestJSON = JSON.parse(rawtext);
+	let tagName = latestJSON.tag_name;
+	if (tagName) {
+		return { succeeded: true, result: tagName };
+	}
+	return { succeeded: false, error: [`Failed to retrieve latest Apache Camel K version tag from : ${latestURL}`] };
+}
+
+function checkKamelCLIVersion() : Promise<string> {
+	return new Promise<string>( async (resolve, reject) => {
+		let kamelLocal = kamelCli.create();
+		await kamelLocal.invoke('version')
+			.then( (rtnValue) => {
+				const strArray = rtnValue.split(' ');
+				const version = strArray[strArray.length - 1].trim();
+				console.log(`Apache Camel K CLI (kamel) version returned: ${version}`);
+				resolve(version);
+				return;
+			}).catch (kamelUnavailableRejection(reject));
+	});
+}
+
+export async function handleChangeRuntimeConfiguration() {
+	let runtimeSetting = config.getKamelRuntimeVersionConfig();
+	let newUpgradeSetting = config.getKamelAutoupgradeConfig();
+	if (newUpgradeSetting === true) {
+		if (runtimeSetting && version.toLowerCase() !== runtimeSetting.toLowerCase()) {
+			extension.shareMessageInMainOutputChannel(`Auto-upgrade setting enabled. Updating to default version ${version} of Apache Camel K CLI`);
+			await config.setKamelRuntimeVersionConfig(version).then ( () => {
+				const msg = `Setting for Apache Camel K runtime version has changed to default ${version}. Please restart the workspace to refresh the Camel K cli.`;
```

Comment:
why do I need to restart the workspace? Why just replacing the kamel binary is not enough?

Reply:
I am working to stick with the Kubernetes approach to installing CLIs here, which means doing it once - at startup - and not afterwards. If we want to change that approach, I think we can - but I think it works the way it is and we can address this in a second PR.